CI: release.yml 의 ocaml/setup-ocaml@v3 가 linux-x64 에서 opam 바이너리 찾기 실패

## 증상

v0.9.1 tag push (2026-04-16) 로 트리거된 release workflow `24480895420` 의 `build (ubuntu-latest, linux-x64)` 잡이 Setup OCaml 단계에서 실패.

에러:
```
##[error]Failed to find opam binary for 'linux' and 'x86_64'. Please check if this combination is supported by opam.
```

`--failed` 로 rerun 했지만 같은 에러로 재실패. 일시적 이슈 아님.

## 영향

v0.9.0 은 정상 릴리즈 완료됐으나 v0.9.1 binary 가 GitHub Releases 에 올라가지 않음. dune-project/opam/CHANGELOG 는 main 에 v0.9.1 로 반영되어 있는데 공식 artifact 가 없는 상태.

## 가설

- `ocaml/setup-ocaml@v3` 가 버전 tag 로 해당 action 의 major track 을 가리키는데, upstream action 이 opam 의 download URL 또는 바이너리 resolution 로직 쪽에 회귀를 넣었을 가능성.
- 또는 ubuntu-latest 이미지 변경에 따른 PATH/architecture 감지 문제.

## 조치 후보

1. `ocaml/setup-ocaml@v3` → 명시적 minor tag (`v3.x.y`) 로 pin
2. 공식 `ocaml/setup-ocaml` 이슈 트래커에서 동일 증상 확인 (릴리즈 노트 + open issues)
3. `runs-on: ubuntu-22.04` 로 downgrade 테스트
4. 필요 시 workflow 에 수동 opam install 단계로 대체

## 재현

- Workflow: `.github/workflows/release.yml`
- Failing job: `build (ubuntu-latest, linux-x64)` → step "Setup OCaml"
- Run URL: https://github.com/jeong-sik/masc-mcp/actions/runs/24480895420

## Related

- v0.9.0 은 2026-04-15 18:52 UTC 에 성공. 사이 기간에 upstream 에서 회귀 유입됐을 가능성.

## Update 2026-04-16 (2nd failure, v0.9.2)

v0.9.2 (tag `v0.9.2`, commit `194698f86`) tag push 후 release workflow `24482345736` 도 **동일 증상으로 실패**:

- `build (ubuntu-latest, linux-x64)` → `##[error]Failed to find opam binary for 'linux' and 'x86_64'`
- `build (macos-14, macos-arm64)` → cancelled (fast-fail)
- `release` → skipped

### 시도된 완화책

#7481 에서 다음 4 ref 를 `ocaml/setup-ocaml@v3` → `@v3.6.0` 로 pin:

```
.github/workflows/release.yml               → @v3.6.0
.github/workflows/webrtc-live-interop.yml   → @v3.6.0
.github/workflows/deploy-railway.yml        → @v3.6.0
.github/actions/setup-ocaml-toolchain/action.yml → @v3.6.0
```

### 의미

Action 자체 (`ocaml/setup-ocaml@v3.6.0`) 는 2026-03-30 에 publish 됐고 그 후 수정 없음. v0.9.0 release 는 2026-04-15 18:52 에 같은 v3.6.0 으로 성공. v0.9.1 은 2026-04-16 early 에 실패. 즉 **action 코드는 불변, 실행 환경(opam binary 호스팅)만 변경**.

**근본 원인 후보**:
1. Upstream opam 바이너리 배포 (e.g. github.com/ocaml/opam releases) 의 asset 명명/위치 변경
2. `ubuntu-latest` image 의 ARCH 감지/리졸버 회귀
3. opam.ocaml.org 호스팅 일시/영구 장애

사용자 repo 측에서 action version pin 만으로는 복구 불가. Upstream (ocaml/opam 또는 ocaml/setup-ocaml) 수정 대기 필요.

### 현재 상태 요약

- **v0.9.2 code**: main 에 반영 완료 (commit `194698f86`).
- **v0.9.2 tag**: push 완료 (`v0.9.2`).
- **v0.9.2 binary release**: **실패** — GitHub Releases 에 asset 없음. 운영자는 v0.9.0 binary 를 계속 사용해야 함.

### 다음 단계

- Upstream (ocaml/setup-ocaml issue tracker) 에 동일 증상 보고가 있는지 확인 → action.yml 또는 opam binary 로드 경로 adjustment
- 원인 해결 전까지 `--no-verify` release 는 하지 않음 (binary 없는 tag 는 오히려 운영 혼란)
- 수정 후 재-tag 전략: v0.9.2 tag 를 유지하면서 release workflow 만 재실행 (`gh workflow run release.yml --ref v0.9.2`) 또는 v0.9.3 bump 를 같이 진행.

--- a/.github/actions/setup-ocaml-toolchain/action.yml
+++ b/.github/actions/setup-ocaml-toolchain/action.yml
@@ -19,7 +19,7 @@ runs:
   using: composite
   steps:
     - name: Setup OCaml
-      uses: ocaml/setup-ocaml@v3.6.0
+      uses: ocaml/setup-ocaml@v3.5.1
       with:
         ocaml-compiler: ${{ inputs.ocaml-compiler }}
         dune-cache: ${{ inputs.dune-cache }}

--- a/.github/workflows/deploy-railway.yml
+++ b/.github/workflows/deploy-railway.yml
@@ -16,7 +16,7 @@ jobs:
       - uses: actions/checkout@v5
 
       - name: Setup OCaml
-        uses: ocaml/setup-ocaml@v3.6.0
+        uses: ocaml/setup-ocaml@v3.5.1
         with:
           ocaml-compiler: 5.4.x
 

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -39,7 +39,7 @@ jobs:
           fi
 
       - name: Setup OCaml
-        uses: ocaml/setup-ocaml@v3.6.0
+        uses: ocaml/setup-ocaml@v3.5.1
         with:
           ocaml-compiler: 5.4.x
 

--- a/.github/workflows/webrtc-live-interop.yml
+++ b/.github/workflows/webrtc-live-interop.yml
@@ -17,7 +17,7 @@ jobs:
         uses: actions/checkout@v5
 
       - name: Setup OCaml
-        uses: ocaml/setup-ocaml@v3.6.0
+        uses: ocaml/setup-ocaml@v3.5.1
         with:
           ocaml-compiler: 5.4.x
           dune-cache: true

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -42,11 +42,12 @@ had the full Phase 2 release to migrate.
 
 ### Fixed
 
-- **Release workflow GitHub Action pinned**. `ocaml/setup-ocaml@v3`
-  floating tag regressed between v0.9.0 (2026-04-15) and v0.9.1
-  (2026-04-16) — opam binary resolution fails on `ubuntu-latest` + `x86_64`.
-  Pinned to `ocaml/setup-ocaml@v3.6.0` (latest release as of 2026-03-30)
-  across `release.yml`, `webrtc-live-interop.yml`, `deploy-railway.yml`.
+- **OCaml setup action pin corrected**. The floating `ocaml/setup-ocaml@v3`
+  tag regressed between v0.9.0 (2026-04-15) and v0.9.1 (2026-04-16), and
+  the follow-up `v3.6.0` pin was not a published setup-ocaml release.
+  Re-pinned the shared setup action plus `release.yml`,
+  `webrtc-live-interop.yml`, and `deploy-railway.yml` to the latest
+  published `ocaml/setup-ocaml@v3.5.1` release verified on 2026-04-16.
   Closes #7475.
 
 ## [0.9.1] - 2026-04-16

